File: frontend/display_app/lib/data/bundled_catalog_repository.dart
import 'dart:convert';

import 'package:flutter/services.dart' show rootBundle;

import '../models/category.dart';
import '../models/product.dart';
import 'backend_dto.dart';
import 'catalog_repository.dart';

/// Fully-offline [CatalogRepository] backed by a bundled snapshot of the real
/// catalog (`assets/catalog_snapshot.json`, exported from the backend). Used in
/// **box-as-server / offline mode**: the display hydrates its catalogue from the
/// bundle with no HTTP and no internet. Placeholder images render on-device.
class BundledCatalogRepository implements CatalogRepository {
  BundledCatalogRepository();

  List<Product>? _cache;

  Future<List<Product>> _ensure() async {
    if (_cache != null) return _cache!;
    final String raw = await rootBundle.loadString(
      'assets/catalog_snapshot.json',
    );
    final Map<String, dynamic> snap = jsonDecode(raw) as Map<String, dynamic>;
    final List<dynamic> products =
        (snap['products'] as List<dynamic>?) ?? const <dynamic>[];
    _cache = products
        .map((dynamic e) => BackendDto.fromDetail(e as Map<String, dynamic>))
        .toList();
    return _cache!;
  }

  @override
  Future<List<Category>> categories() async {
    final List<Product> list = await _ensure();
    final Set<String> seen = <String>{};
    final List<Category> cats = <Category>[];
    for (final Product p in list) {
      if (p.categoryId.isNotEmpty && seen.add(p.categoryId)) {
        cats.add(Category(id: p.categoryId, name: p.categoryId));
      }
    }
    cats.sort((Category a, Category b) => a.name.compareTo(b.name));
    return cats;
  }

  @override
  Future<List<Product>> products({String? categoryId, String? query}) async {
    final List<Product> list = await _ensure();
    final String? q = query?.trim().toLowerCase();
    return list.where((Product p) {
      final bool okCat = categoryId == null || p.categoryId == categoryId;
      final bool okQ =
          q == null ||
          q.isEmpty ||
          p.name.toLowerCase().contains(q) ||
          p.brand.toLowerCase().contains(q) ||
          p.aiHighlights.any((String h) => h.toLowerCase().contains(q));
      return okCat && okQ;
    }).toList();
  }

  @override
  Future<Product?> productById(String id) async {
    final List<Product> list = await _ensure();
    return list.where((Product p) => p.id == id).firstOrNull;
  }
}

File: frontend-rn/display_app/src/data/bundledCatalogRepository.ts
import { Category } from '../models/category';
import { Product } from '../models/product';
import { BackendDto } from './backendDto';
import { CatalogRepository } from './catalogRepository';
// Metro bundles the JSON snapshot directly — no network, no filesystem.
import snapshot from '../../assets/catalog_snapshot.json';

/// Fully-offline CatalogRepository backed by a bundled snapshot of the real
/// catalog (exported from the backend). Ported from `BundledCatalogRepository`.
export class BundledCatalogRepository implements CatalogRepository {
  private cache: Product[] | null = null;

  private ensure(): Product[] {
    if (this.cache != null) return this.cache;
    const products = ((snapshot as any).products as any[]) ?? [];
    this.cache = products.map((e) => BackendDto.fromDetail(e));
    return this.cache;
  }

  async categories(): Promise<Category[]> {
    const list = this.ensure();
    const seen = new Set<string>();
    const cats: Category[] = [];
    for (const p of list) {
      if (p.categoryId.length > 0 && !seen.has(p.categoryId)) {
        seen.add(p.categoryId);
        cats.push(new Category({ id: p.categoryId, name: p.categoryId }));
      }
    }
    cats.sort((a, b) => a.name.localeCompare(b.name));
    return cats;
  }

  async products(opts?: { categoryId?: string; query?: string }): Promise<Product[]> {
    const list = this.ensure();
    const q = opts?.query?.trim().toLowerCase();
    return list.filter((p) => {
      const okCat = opts?.categoryId == null || p.categoryId === opts.categoryId;
      const okQ =
        q == null ||
        q.length === 0 ||
        p.name.toLowerCase().includes(q) ||
        p.brand.toLowerCase().includes(q) ||
        p.aiHighlights.some((h) => h.toLowerCase().includes(q));
      return okCat && okQ;
    });
  }

  async productById(id: string): Promise<Product | undefined> {
    const list = this.ensure();
    return list.find((p) => p.id === id);
  }
}
